MoreLinq/DeconstructibleEnumerable.cs

```diff
@@ -0,0 +1,86 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2019 Atif Aziz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+
+    /// <summary>
+    /// Represents a source whose elements can be sequentially deconstructed
+    /// into separate variables.
+    /// </summary>
+    /// <typeparam name="T">The type of the elements.</typeparam>
+
+    public partial struct DeconstructibleEnumerable<T> : IEnumerable<T>
```

Comment:
I assume this is a struct for optimization purposes. I wonder if we are not painting ourselves into a corner by restricting us to a single implementation. I could imagine wanting to implement a specialization for sources implementing `IList<T>`. OTOH, given that the number of elements is always short, maybe that optimization won't ever matter. What do you think?

Reply:
The `struct` is there to make the _explicit intention to deconstruct_ cheap. Optimizations for other collection types would still be possible but they would belong with the `Deconstruct` overloads as opposed to `AsDeconstructible` deciding which implementation to return based on the source type. In other words, this is how I see it:

```c#
public void Deconstruct(out T item1, out T item2)
{
    const int count = 2;

    if (_source is List<T> list)
    {
        if (list.Count != 2)
            throw list.Count < count ? CreateTooShortError(count) : CreateTooLongError(count);
        item1 = list[0];
        item2 = list[1];
    }
    else
    {
        using (var e = GetEnumerator())
        {
            item1 = e.MoveNext() ? e.Current : throw CreateTooShortError(count);
            item2 = e.MoveNext() ? e.Current : throw CreateTooShortError(count);
            if (e.MoveNext())
                throw CreateTooLongError(count);
        }
    }
}
```

The decision to optimize is therefore deferred and delegated. The only downside I see to this is that the type test to choose an optimization path would have to be done on each call to `Deconstruct`, though that said, I honestly don't see anyone deconstructing more than once for any practical purpose as it would be costly unless the source is materialized or memoized.

Reply:
@fsateler Thoughts on this?

Reply:
I'm still giving it some thought. My instincts is towards the interface, but I don't think that practically we will need to ever add that optimization. Given the low item count, I'd be surprised if there was a significant gain to be made.